batch/batch/driver/main.py

```diff
@@ -1249,15 +1254,16 @@ async def cancel_fast_failing_batches(app):
 
     records = db.select_and_fetchall(
         """
-SELECT batches.id, job_groups_n_jobs_in_complete_states.n_failed
-FROM batches
+SELECT job_groups.batch_id, job_groups_n_jobs_in_complete_states.n_failed
+FROM job_groups
 LEFT JOIN job_groups_n_jobs_in_complete_states
-  ON batches.id = job_groups_n_jobs_in_complete_states.id
-WHERE state = 'running' AND cancel_after_n_failures IS NOT NULL AND n_failed >= cancel_after_n_failures
-"""
+  ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
+WHERE state = 'running' AND cancel_after_n_failures IS NOT NULL AND n_failed >= cancel_after_n_failures AND job_groups.job_group_id = %s
+""",
+        (ROOT_JOB_GROUP_ID,),
```

Comment:
What is this function for? Why isn't it in the canceller? Why does it only apply to the root job group?

Reply:
This is the functionality that polls and looks for batches / job groups with more than cancel_after_n_failures and then cancels those batches / job groups in the database. The canceller actually iterates through the jobs and cancels them.